core/optaplanner-constraint-streams-bavet/src/main/java/org/optaplanner/constraint/streams/bavet/common/index/ComparisonIndexer.java

```diff
@@ -119,14 +124,30 @@ public boolean isEmpty() {
 
     private static final class KeyComparator<Key_ extends Comparable<Key_>> implements Comparator<Key_> {
 
-        private static final Comparator COMPARATOR = new KeyComparator<>();
-        private static final Comparator COMPARATOR_REVERSED = COMPARATOR.reversed();
+        private static final Comparator INSTANCE = new KeyComparator<>();
 
         @Override
-        public int compare(Key_ o1, Key_ o2) { // Exists so that the comparison operations can be more easily debugged.
+        public int compare(Key_ o1, Key_ o2) {
+            if (o1 == o2) {
+                return 0;
+            }
             return o1.compareTo(o2);
         }
 
     }
 
+    private static final class ReversedKeyComparator<Key_ extends Comparable<Key_>> implements Comparator<Key_> {
+
+        private static final Comparator INSTANCE = new ReversedKeyComparator<>();
```

Comment:
Can we use `Collections.reverseOrder()`?

If available in Java 11 already. And if not, add a TODO to delete our own impl when we upgrade to java 17?